### DISCUSSION TITLE
Ideas / wishes for Nitro v3

### DISCUSSION THREAD

**Discussion by MickL** [nitrojs/nitro#2385] [Ideas]
You [asked](https://github.com/unjs/nitro/discussions/2324) to start a new discussion if we have ideas for v3, so here are my wishes:

- Add testing utilities (bootstrap, env, fetch, mocking) #2303 
- Totally remove auto imports and make it an opt-in #2232 #2007
- Add missing pieces to docs, not everything is documented e.g. modules #1862
- Add more examples to the docs (compare docs to NestJS, AdonisJS, Angular), e.g. session auth with h3 useSession #2303

**Comment by InsaneNaman** [comment]
I have been tracking this project since last vite conf. I feel this project has lot of potential but I would like to politely mention that it needs extensive docs. Many devs are not even aware of this great project.

**Reply by MickL** [reply]
I fully agree about the docs

**Reply by vladkrasn** [reply]
Yeah, as a backend engineer I am interested in Nitro, but the lack of proper documentation prevents me from even trying it.
I don't know what are the actual capabilities of this project; I don't know if it's supposed to be used for building complex backends or not; Many features, I suspect, are either undocumented or documented in other packages - like h3, db0 or maybe even Nuxt (and I don't have any experience with Nuxt, unlike many people in the unjs commiunity)

**Reply by MickL** [reply]
I think I have most of my knowledge from Nuxt. I already noted that it is a problem that documentation is split into H3, Nitro and Nuxt while most is in Nuxt actually.

Additionally the ecosystem of Nitro is non-existent. It is possible to write modules (similar to Nuxt modules) but this feature is undocumented: #1862 . Further modules are developed for Nuxt only and Nitro gets nothing from the cake. For example Nuxt team is working on nuxt-auth but I dont expect nitro-auth to be coming. Or there is nuxt-security but not nitro-security: https://github.com/Baroshem/nuxt-security/issues/405

I dont think Nitro is missing anything from bigger frameworks (except for testing capabilities) like NestJS or AdonisJS: We have caching, validation, security and everything else can be implemented e.g. auth with Lucia AND Nitro can be deployed everywhere which the big ones cant, BUT what is missing is documentation and tutorials.

**Reply by InsaneNaman** [reply]
I really appreciate the efforts by the nitro team and feel the same. My expertise lies in frontend (React) and unaware about the Nuxt ecosystem in depth but nitro got my attention and still i am also on the same page i.e, I dont know what nitro wants it to be (A backend framework or framework for framework or both). I am for now going towards Hono but really wants to see nitro catching up.

**Reply by ivanjeremic** [reply]
> For example Nuxt team is working on nuxt-auth but I don't expect nitro-auth to be coming. Or there is nuxt-security but not nitro-security

Very sad to hear that the plugins are not compatible even though Nuxt is using Nitro under the hood if I understood correctly, the ecosystem would benefit if these parts were also developed in an agnostic way, and that we may get in the future things like `unauth` & `unsecurity`

**Reply by MickL** [reply]
I dont know if it will be not compatible but I expect it to be a Nuxt exclusive module. My wish would be to build modules for Nitro first and then release the same for Nuxt if necessary.

**Reply by ivanjeremic** [reply]
> My wish would be to build modules for Nitro first and then release the same for Nuxt if necessary.

That sounds like the most logical and clean way to do things, that way other projects outside of Nuxt can benefit from it which would also align more with the UNJS tagline on the website `Agnostic Excellence: JavaScript Libraries, Tools, and Utilities, Crafted to Elevate Your Coding Journey.`

**Reply by InsaneNaman** [reply]
I agree with @MickL and @ivanjeremic. If nitro becomes this then it will be a go to laravel of node. I know adonis is there but idk mvc doesn't excites me that much. Nitro has the simplicity, better ecosystem, big projects already using it (nuxt) etc. Make it happen team nitro xD

**Reply by pi0** [reply]
Thanks for the kind words! We will definitely get there sometime (I hope soon, maybe this year), Nitro surely deserves a more mature ecosystem. As the creator of both Nuxt modules and Nitro, I know the value of Modules in the expansion of the ecosystem but at the same time, the fragmentation it can cause in not mature state. Every step we take at unjs is in the goodwill of both direct users and ecosystem health + adoption which, i admit makes them much much harder decisions that can be made at meta-framework level but are certainly doable and worth waiting.

**Reply by MickL** [reply]
I worked years with NestJS and I tried AdonisJS and to me Nitro can be the next big framework especially because its light weight and simple and to me and my customer a "Node" api that can run serverless is a big thing. The only thing missing are testing utilities and extensive docs and articles about it.

**Reply by MickL** [reply]
I just found out by accident that H3 (the server Nitro uses) [can handle sessions](https://h3.unjs.io/examples/handle-session) and therefor there is no need for an auth library like Lucia that handles session management. The only thing missing is documentation or an article/tutorial about it ... I am in the Nitro docs and see zero results for the search term "sessions" and of course no article/tutorial can be found for that topic. For Nuxt there is an [example project](https://github.com/nuxt/examples/tree/main/examples/auth/local) that can be used nearly 1 to 1 for Nitro, but whoever is using Nitro only will never find it.

**Reply by ivanjeremic** [reply]
> I just found out by accident that H3 (the server Nitro uses) can handle sessions and therefor there is no need for an auth library like Lucia that handles session management. 

I'm in the middle of integrating Lucia and kind of shocked now, I think it is still the best option and the session features can be combined to work together somehow, if someone has a better library please suggest it. Really hope the Docs catch up soon.

**Comment by alumni** [comment]
1. REST API generation: I need to generate a REST API for about a hundred entities. It would be nice to have a plug in that does this automatically.
2. Dependency injection: I have a lot of server-side code which  would look ugly organized in a functional way. It would be nice to be able to create class-based route handlers (controllers) which would have services injected. It would be nice to override the injected service when testing (or when running locally).

**Reply by qodeboy** [reply]
Have you managed to find a nice way around #2?

**Reply by MickL** [reply]
I come from a framework that uses DI (NestJS) and I feel like not having this is a GREAT relief. But this is only due to what I learned within Nuxt, how to create reusable utilities and composables. So what I want to say is that with a different style of programming everything can become much cleaner and easier and DI is no more needed. For people that never worked with Nuxt before this concepts should be explained in the Nitro docs imo.

What I am missing tho is also the possibility to overwrite/mock stuff for dev or testing. 

Btw for #1 you could write your own plugin maybe?

**Reply by alumni** [reply]
I did not bother to try nitro/nuxt more than that. I just looked at quite a few sample repositories, did not find anything interesting, moved on.

Honestly, SAP CAP is doing what nuxt is doing and whole bunch more, so if I'd switch to something, I'd rather switch to that.

However, I stuck with NestJS, I like the design pattern more, it's more intuitive.
